views/clickhouse.go

```diff
@@ -0,0 +1,151 @@
+
+type RealtimeViewershipRow struct {
+	Timestamp   time.Time `ch:"timestamp_ts"`
+	ViewCount   uint64    `ch:"view_count"`
+	BufferRatio float64   `ch:"buffer_ratio"`
+	ErrorRate   float64   `ch:"error_rate"`
+
+	PlaybackID  string `ch:"playback_id"`
+	DeviceType  string `ch:"device_type"`
+	Browser     string `ch:"browser"`
+	CountryName string `ch:"playback_country_name"`
+}
+
+type Clickhouse interface {
+	QueryRealtimeViewsEvents(ctx context.Context, spec QuerySpec) ([]RealtimeViewershipRow, error)
+}
+
+type ClickhouseOptions struct {
+	Addr     string
+	User     string
+	Password string
+	Database string
+}
+
+type ClickhouseClient struct {
+	conn driver.Conn
+}
+
+func NewClickhouseConn(opts ClickhouseOptions) (*ClickhouseClient, error) {
+	conn, err := clickhouse.Open(&clickhouse.Options{
+		Addr: strings.Split(opts.Addr, ","),
+		Auth: clickhouse.Auth{
+			Database: opts.Database,
+			Username: opts.User,
+			Password: opts.Password,
+		},
+		TLS: &tls.Config{},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &ClickhouseClient{conn: conn}, nil
+}
+
+func (c *ClickhouseClient) QueryRealtimeViewsEvents(ctx context.Context, spec QuerySpec) ([]RealtimeViewershipRow, error) {
+	sql, args, err := buildRealtimeViewsEventsQuery(spec)
+	if err != nil {
+		return nil, fmt.Errorf("error building viewership events query: %w", err)
+	}
+	var res []RealtimeViewershipRow
+	err = c.conn.Select(ctx, &res, sql, args...)
+	if err != nil {
+		return nil, err
+	}
+	res = replaceNaNBufferRatio(res)
+
+	return res, nil
+}
+
+func buildRealtimeViewsEventsQuery(spec QuerySpec) (string, []interface{}, error) {
+	var query squirrel.SelectBuilder
+	if spec.From == nil && spec.To == nil {
+		query = currentEventsQuery(spec)
+	} else {
+		query = timeRangeEventsQuery(spec)
+	}
+
+	query = withPlaybackIdFilter(query, spec.Filter.PlaybackID)
+	if creatorId := spec.Filter.CreatorID; creatorId != "" {
+		query = query.Where("creator_id_type = ?", "unverified")
+		query = query.Where("creator_id = ?", creatorId)
+	}
+
+	for _, by := range spec.BreakdownBy {
+		field, ok := realtimeViewershipBreakdownFields[by]
+		if !ok {
+			return "", nil, fmt.Errorf("invalid breakdown field: %s", by)
+		}
+		// skip breakdowns that are already in the query
+		// only happens when playbackId or dStorageUrl is specified
+		if sql, _, _ := query.ToSql(); strings.Contains(sql, field) {
+			continue
+		}
+		query = query.Columns(field).GroupBy(field)
+	}
+
+	sql, args, err := query.ToSql()
+	if err != nil {
+		return "", nil, err
+	}
+
+	return sql, args, nil
+}
+
+// currentEventsQuery uses an unoptimized raw playback logs to favor latency over query speed.
+func currentEventsQuery(spec QuerySpec) squirrel.SelectBuilder {
+	return squirrel.Select(
+		"count(distinct session_id) as view_count").
+		From("viewership_events").
+		Where("user_id = ?", spec.Filter.UserID).
+		Where("server_timestamp > (toUnixTimestamp(now() - 30)) * 1000").
+		Limit(maxClickhouseResultRows)
```

Comment:
Updated in c3a1e2d684f3bd448e835122550012afc66115fb

Though, in the case of Clickhouse and the realtime viewership API it's slightly different because we don't allow specifying `timeStep` and the time rage is max 3h. Then, the max number of rows is `3 * 60 * dimension`. Still, if someone specifies a lot of dimensions we may hit the limit.